### ISSUE TITLE
[blog post] On npm's node_module resolution

### ISSUE THREAD

**Issue by dev-cprice** [dev-cprice/codyaprice.com#85]
How does npm module resolution work?

Anonymized discussion from work below:

> Caveats and nuance do apply.
>
> Node's module resolution (mostly their hoisting algorithm) can really bite you if you're not careful.
> 
> We used to put a list of dependencies (read: not `devDependencies` or `peerDependencies`) in the work-specific SDK. We don't do this anymore. Our original rationale was that there were a lot of common things that all consumers used and we wanted to abstract that away from them and reduce the barrier of entry for developing front end apps. This abstraction not only caused confusion because people no longer knew where certain packages were coming from, but it also caused the wrong versions of packages to resolve in apps. The source of truth is what's in your app's package.json, not what's in a dependency's. The setup was something like the following:
>
> work specific SDK's package.json
>```json
>{
>  "name": "work-specific-sdk",
>  "version": "1.0.0",
>  "dependencies": {
>    "react": "^16.0.0",
>    "terra-button": "^3.0.0"
>  }
>}
>```
>
> your app's package.json:
>```json
>{
>  "name": "my-app",
>  "dependencies": {
>    "work-specific-sdk": "^1.0.0",
>    "foo-dep": "^3.0.0"
>  }
>}
>```
>
> foo-dep's package.json
>```json
>{
>  "name": "foo-dep",
>  "version": "3.0.0",
>  "dependencies": {
>    "terra-button": "^2.0.0"
>  }
>}
>```
>
>In this scenario, when your app imports `terra-button`, which version should you get? Which version do you think npm hoisted to the root level that your app includes?
>
>The answer: it's indeterminate which version you'll get. Usually, npm chose to resolve your app's import to the lower version. So, you would think and expect to import the latest and greatest version of `terra-button` and from the work-specific-sdk, but you would not get it.
>
>TLDR: proceed with caution, def dragons. 🐉
